link-grammar/parse/parse.c

```diff
@@ -247,6 +246,55 @@ static void sort_linkages(Sentence sent, Parse_Options opts)
 	print_time(opts, "Sorted all linkages");
 }
 
+#define SHORTEST_SENTENCE_TO_PACK 9
+
+/**
+ * Pack all disjunct and connectors into one big memory block.
+ * This facilitate a better memory caching for long sentences
+ * using the fact that now more than one connector can be packed
+ * in a cache line of 64 bytes (a performance gain of a few percents).
+ *
+ * FIXME: 1. Find the "best" value for SHORTEST_SENTENCE_TO_PACK.
+ * 2. Maybe this check should be done in too stages, the second one
+ * will use number of disjunct and connector thresholds.
+ */
+static void pack_sentence(Sentence sent)
+{
+	int dcnt = 0;
+	int ccnt = 0;
+
+	if (sent->length < SHORTEST_SENTENCE_TO_PACK) return;
+	for (size_t w = 0; w < sent->length; w++)
+	{
+		Disjunct *d;
+
+		for (d = sent->word[w].d; NULL != d; d = d->next)
+		{
+			dcnt++;
+			for (Connector *c = d->right; c!=NULL; c = c->next) ccnt++;
+			for (Connector *c = d->left; c != NULL; c = c->next) ccnt++;
+		}
+	}
+
+#define CACHELINE 64
+	size_t dsize = dcnt * sizeof(Disjunct);
+	dsize = (dsize+CACHELINE)&~(CACHELINE-1); /* Align connector block. */
```

Comment:
Here, you align, but then below, you add csize of unknown alignment.   I'm sort-of confused, because you then use what malloc provided, without actually aligning -- malloc might provide an address that is not 64-byte aligned.

Reply:
The memory arrangement in this allocation is:
DISJUNCT_ARRAY **ALIGNEMNT_PAD** CONNECTOR_ARRAY
The Connector struct is 32 bytes (and can be shrinked to 16, 8 and even 4 bytes - I would like to test all).
In order not to split connectors over two cache lines, I want CONNECTOR_ARRAY to start at 64-byte boundary. I get this here by adding an alignment pad at the **end** of the disjunct allocation.
(As far as I remember the Disjunct struc size is something like 44 bytes so there is no point to align its start and it doesn't usually end at the desired boundary.)

So it seems fine to me.
If desired, I can add a description like the above to the function comment.

Reply:
> As far as I remember the Disjunct struc size is something like 44

(It is actually 56.)

Reply:
OK. If the code likes right to you, then OK, it looked strange to me but I did not try very hard to figure it uout.

Reply:
I hope it will be clearer after I add comments to it.